File: python_version/binary_tree/path-sum ii.py
class Solution:
    def pathSum(self, root, sum):
        if not root:
            return []
        result = []
        self.helper(root,sum,0,[],result)
        return result

    def helper(self,root,sum,temp,path,result):
        if not root:
            return
        temp += root.val
        path.append(root.val)
        if not root.left and not root.right:
            if sum == temp:
                result.append(path)
        self.helper(root.left, sum, temp, [i for i in path],result)
        self.helper(root.right,sum, temp, [i for i in path],result)

        

File: src/dfs/path-sum-ii.java
package dfs;

import java.util.*;

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode(int x) {
    val = x;
  }
};

class FindAllTreePaths {
    public static List<List<Integer>> findPaths(TreeNode root, int sum) {
      List<List<Integer>> allPaths = new ArrayList<>();
      List<Integer> current = new ArrayList<>();
  
      dfs(allPaths, current, root, sum);
      return allPaths;
    }
  
    public static void dfs(List<List<Integer>> allPaths, List<Integer> currentPath, TreeNode currentNode, int sum){
      if(currentNode == null){
        return;
      }
      currentPath.add(currentNode.val);
  
      if(currentNode.val == sum && currentNode.left == null && currentNode.right == null){
        allPaths.add(new ArrayList<>(currentPath));
      }else{
        dfs(allPaths, currentPath, currentNode.left, sum - currentNode.val);
        dfs(allPaths, currentPath, currentNode.right, sum - currentNode.val);
      }
  
      currentPath.remove(currentPath.size() - 1);
    }
}